### PR TITLE
[pro#208] Limit embargo extensions

### PR DESCRIPTION
**Pull request by lizconlan**
Closes mysociety/alaveteli-professional#208

Prevents requests under embargo from receiving an embargo extension unless the existing embargo is nearing expiry. Currently places admins under the same restriction which is the easiest thing to do but may prove problematic (the restriction applies at view and controller levels but there are no such restriction on the models so it can be circumvented on the console if required).

### CODE REVIEW

**Comment by lizconlan** [comment]
Hmm, I have subsequently realised there is already an `embargo_expiring?` method on `InfoRequest` which I could have used. However I quite like having the extra check for the expiration date having passed a (presumably rare) edge case outside the test suite when the date has rolled over but the embargo has not yet been removed by the task runner.

File: app/models/alaveteli_pro/embargo.rb [hunk 1]
```diff
@@ -71,6 +71,11 @@ def extend(extension)
       save
     end
 
+    def expiring_soon?
```
[the remark this answers is not in the document]
Comment (lizconlan) [reply app/models/alaveteli_pro/embargo.rb]: Hmm, despite coming to the conclusion that "is the publication date less than a week away?" (`InfoRequest#embargo_expiring?`) is essentially the same question as "is the current time between a week from publication and the publication deadline?" (`Embargo#expiring_soon?`), I've added an extra check to `embargo_expiring?` to handle possible edge cases where the intention is that the embargo has expired but for some reason the embargo record has not been removed - in the case of it being really early on the publication day, that could be a matter of timing, after that date presumably some kind of error

Will happily stand corrected if this does not sound right

File: app/views/alaveteli_pro/info_requests/_embargo_form.html.erb [hunk 2]
```diff
@@ -11,27 +11,29 @@
...
   <input class="houdini-input" type="checkbox" id="input1">
   <div class="houdini-target extend-embargo-sidebar">
-    <%= form_for(
-          [AlaveteliPro::EmbargoExtension.new(embargo: info_request.embargo)],
-          html: {class: 'js-embargo-form'}) do |f| %>
-      <%= render partial: "alaveteli_pro/info_requests/embargo_info",
+    <%= render partial: "alaveteli_pro/info_requests/embargo_info",
                  locals: {embargo: info_request.embargo, tense: :present} %>
-      <%= f.hidden_field :embargo_id %>
+    <% if info_request.embargo.expiring_soon? %>
```
[the remark this answers is not in the document]
Comment (lizconlan) [reply app/views/alaveteli_pro/info_requests/_embargo_form.html.erb]: Adding a message that indicates when they will be able to update it as that feels less frustrating

File: app/views/alaveteli_pro/info_requests/_embargo_form.html.erb [hunk 3]
```diff
@@ -34,6 +34,13 @@
                  class="embargo__submit js-embargo-submit">
         </p>
       <% end %>
+    <% else %>
+      <p>
+        <%= _("If you need to keep your request private for longer, you will " \
+              "be able to extend this privacy period on " \
+              "{{embargo_extend_from}}",
```
Comment (garethrees) [inline app/views/alaveteli_pro/info_requests/_embargo_form.html.erb]: Maybe make this a bit more concise?

```erb
<%= _("You will be able to extend this privacy period from {{embargo_extend_from}}.",
      embargo_extend_from: embargo_extend_from(info_request.embargo)) %>
```

File: app/models/info_request.rb [hunk 4]
```diff
@@ -1577,15 +1577,12 @@ def self.log_very_overdue_events
...
+  # has already occurred but the embargo has not yet been deleted)
   #
   # Returns boolean
   def embargo_expiring?
-    if self.embargo
-      self.embargo.publish_at <= AlaveteliPro::Embargo.expiring_soon_time
-    else
-      false
-    end
+    !!embargo && (embargo.expiring_soon? || Time.zone.now >= embargo.publish_at)
```
[the remark this answers is not in the document]
Comment (lizconlan) [reply app/models/info_request.rb]: no, wait - the opposite of that, we want `embargo_ending_expiry?`

File: app/models/info_request.rb [hunk 5]
```diff
@@ -1584,6 +1584,13 @@ def embargo_expiring?
     embargo ? embargo.expiring_soon? : false
   end
 
+  # Is the attached embargo still present but has reached its publication date
+  #
+  # Returns boolean
+  def embargo_pending_expiry?
```
Comment (garethrees) [inline app/models/info_request.rb]: Its not the end of the world, but not sure what the differing method name adds here. `embargo_expired?` is a bit clearer in that its more obviously delegating to `embargo.expired?`.

File: spec/models/info_request_spec.rb [hunk 6]
```diff
@@ -3515,6 +3515,48 @@ def email_and_raw_email(opts = {})
...
+          expect(info_request.reload.embargo_pending_expiry?).to be true
+        end
+      end
+
+    end
+
+    context 'when there is no embargo' do
+
+      it 'returns true' do
+        expect(info_request.embargo_pending_expiry?).to be false
```
Comment (garethrees) [inline spec/models/info_request_spec.rb]: The spec description and assertion differ here.

Comment (lizconlan) [reply spec/models/info_request_spec.rb]: /o\ (thank you)

File: app/models/info_request.rb (hunk 5 above)
Comment (lizconlan) [reply app/models/info_request.rb]: there's a different (and more accurate) `embargo_expired?` [over in pro#208](https://github.com/mysociety/alaveteli/pull/4296/files#diff-5f7dfe1185cac97cbc9f9dc356abce7cR1583) which is intended to only return true if the request has been published (and there was once an embargo), happy to come up with a better name here though - this describes an edge case which hopefully will only be true between midnight and the scheduled script run which deletes pending embargoes and causes the requests to be reindexed

Comment (garethrees) [reply app/models/info_request.rb]: Ah right yeah, ISWYM.

I think this is fine in that case, and we can look at the `embargo_*` method names together when all merged to check each makes sense in context.

Comment (lizconlan) [reply app/models/info_request.rb]: is it worth me combining the 3 overlapping PRs into one once I've done the fixups so that we can review/check pre-merge?
